pgx/src/datum/sql_entity_graph/mod.rs

```diff
@@ -64,6 +64,72 @@ pub trait ToSql {
     fn to_sql(&self, context: &PgxSql) -> eyre::Result<String>;
 }
 
+/// Represents a pointer to a function that can transform a SqlGraphEntity to a SQL string
+///
+/// This is used to provide a facility for overriding the default SQL generator behavior using
+/// the `#[to_sql(path::to::function)]` attribute in circumstances where the default behavior is
+/// not desirable.
+///
+/// Implementations can invoke `ToSql::to_sql(entity, context)` on the unwrapped SqlGraphEntity
+/// type should they wish to delegate to the default behavior for any reason.
+pub type ToSqlFn = fn(&SqlGraphEntity, &PgxSql) -> eyre::Result<String>;
+
+/// Represents configuration options for tuning the SQL generator.
+///
+/// When an item that can be rendered to SQL has these options at hand, they should be
+/// respected. If an item does not have them, then it is not expected that the SQL generation
+/// for those items can be modified.
+///
+/// The default configuration has `enabled` set to `true`, and `callback` to `None`, which indicates
+/// that the default SQL generation behavior will be used. These are intended to be mutually exclusive
+/// options, so `callback` should only be set if generation is enabled.
+///
+/// When `enabled` is false, no SQL is generated for the item being configured.
+///
+/// When `callback` has a value, the corresponding `ToSql` implementation should invoke the
+/// callback instead of performing their default behavior.
+#[derive(Clone)]
+pub struct ToSqlConfig {
+    pub enabled: bool,
+    pub callback: Option<ToSqlFn>,
+}
+impl std::cmp::PartialEq for ToSqlConfig {
+    fn eq(&self, other: &Self) -> bool {
+        if self.enabled != other.enabled {
+            return false;
+        }
+        match (self.callback, other.callback) {
+            (None, None) => true,
+            (Some(a), Some(b)) => std::ptr::eq(std::ptr::addr_of!(a), std::ptr::addr_of!(b)),
+            _ => false,
+        }
+    }
+}
+impl std::cmp::Eq for ToSqlConfig {}
+impl std::hash::Hash for ToSqlConfig {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        self.enabled.hash(state);
+        self.callback.map(|cb| std::ptr::addr_of!(cb)).hash(state);
+    }
+}
+impl std::fmt::Debug for ToSqlConfig {
```

Comment:
This couldn't be derived? Interesting.

Reply:
Yeah, basically rustc complains that it can't derive them because in order for the derivation to be valid, it would have to be implemented for all possible lifetimes that the references in the argument list could be valid for, e.g. you get errors like this:

```
error: implementation of `Debug` is not general enough
  --> pgx/src/datum/sql_entity_graph/mod.rs:94:5
   |
91 | #[derive(Debug, Clone)]
   |          ----- in this derive macro expansion
...
94 |     pub callback: Option<ToSqlFn>,
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ implementation of `Debug` is not general enough
   |
   = note: `Debug` would have to be implemented for the type `for<'r, 's> fn(&'r SqlGraphEntity, &'s PgxSql) -> std::result::Result<std::string::String, ErrReport>`
   = note: ...but `Debug` is actually implemented for the type `fn(&'0 SqlGraphEntity, &PgxSql) -> std::result::Result<std::string::String, ErrReport>`, for some specific lifetime `'0`
   = note: this error originates in the derive macro `Debug` (in Nightly builds, run with -Z macro-backtrace for more info)
```

So the workaround here is just formatting as a raw pointer with no associated lifetime. The other handwritten impls are necessary for similar reasons (either there is no derivable impl, or the derivation has to be done by hand because of the lifetime issue).

Reply:
Whoa, super interesting. Thanks for trying this. :)